### ISSUE TITLE
SBT plugin is looking for a new maintainer

### ISSUE THREAD

**Issue by olafurpg** [olafurpg/scalafmt#597]
I have decided to remove the SBT plugin. Why? I don't personally use the SBT plugin, the plugin has a few [issues](https://github.com/olafurpg/scalafmt/labels/sbt) and maintaining the plugin in my free time has turned out to be an unrewarding experience. I have several times asked for help to maintain the plugin, but unfortunately with no luck so far.

It appears many people (and organizations) use the sbt plugin. I'm opening this issue to invite anyone who is interested to lend an extra hand to help bring the SBT plugin back. I would love to welcome you as a collaborator to the project, give you merge rights, and help bring you up to speed on making the SBT plugin a success.

To show your interest in maintaining the SBT plugin, you can do any of the following

- send a PR reverting 308f393ac71cbb290bc8166c9f82bed59b1f3a17 and 983b04372b76fd116c1b3767ac78012367412f44. This is probably the easiest option, but comes with the baggage of issues such as #485, which is caused by a [regression in sbt 0.13.12](https://github.com/sbt/sbt/issues/2786). The implementation is mostly borrowed from the scalariform sbt plugin but is slightly more complicated because scalafmt is 2.11 only and sbt plugins must compile against 2.10.
- send a PR implementing a new SBT plugin from scratch that uses the [`bootstrap` module](https://github.com/olafurpg/scalafmt/blob/master/bootstrap/src/main/scala/org/scalafmt/bootstrap/Bootstrap.scala). The bootstrap module can potentially fix issues such as #485 by using the coursier api as a replacement for SBT's `update` mechanism (the root cause of #485). Don't hesitate to ask for more details.
- send a PR creating a new SBT plugin that leverages the ["synthetic projects" feature in sbt 0.13.13](https://github.com/sbt/sbt/releases/tag/v0.13.13) to create an isolated classpath to invoke the scalafmt CLI
- send a PR implementing a new SBT plugin with whatever approach you like. However, please note that sbt plugins must compile against 2.10 and scalafmt is 2.11 only because of its dependency on scala.meta. Backporting scala.meta to 2.10 has so far been unsuccessful, https://github.com/scalameta/scalameta/issues/295

If you opt to implement a new SBT plugin, please include scripted tests like the previous plugin had.

### Migrating away from the SBT plugin

Use the CLI: https://olafurpg.github.io/scalafmt/#CLI

To enforce scalafmt in a project like was possible with the sbt plugin, my personal recommendation is to

- put a tiny `scalafmt` bootstrap script into your repo as explained [here](https://olafurpg.github.io/scalafmt/#Coursier), 
- configure the `project` setting in .scalafmt.conf as explained [here](https://olafurpg.github.io/scalafmt/#project),
- run `./scalafmt` instead of `sbt scalafmt` and
- run `./scalafmt --test` instead of `sbt scalafmtTest`.
- In CI, you can cache the downloaded jars in `$HOME/.coursier/bootstrap`.
- It takes a few seconds to create a bootstrap script for a new scalafmt version, as long as you have [`coursier`](https://github.com/alexarchambault/coursier#command-line-1) installed.

For your convenience, here is a bootstrap script for scalafmt 0.4.10: https://github.com/olafurpg/scalafmt/files/601602/scalafmt.zip

**Comment by RomanIakovlev** [comment]
I wonder if it's a good idea to combine coursier-style fetching and launching the scalafmt artifact with sbt plugin. I mean, one can write a sbt plugin that uses the coursier to fetch and run scalafmt in the context of particular sbt project. Scalafmt will be launched as a separate process, so all issues with classpaths will go away, but it will still enjoy the sbt integration. I could give it a try if you think it makes sense.

**Comment by olafurpg** [comment]
@RomanIakovlev I think that approach could work. The [bootstrap module](https://github.com/olafurpg/scalafmt/blob/master/bootstrap/src/main/scala/org/scalafmt/bootstrap/Bootstrap.scala) already implements most of the coursier logic to classload a specific version of scalafmt and inject the correct in/out/err streams. See the example test here: https://github.com/olafurpg/scalafmt/blob/master/bootstrap/src/test/scala/org/scalafmt/bootstrap/BootstrapTest.scala

The bootstrap module should be able to cross-compile to 2.10 and 2.11 so you should be able to call it from an sbt plugin. Please give it a try if you like :)

**Comment by olafurpg** [comment]
You are of course also free to roll your own implementation of using the coursier api to invoke scalafmt, whatever you prefer!

**Comment by olafurpg** [comment]
@RomanIakovlev You might also be interested in the approach explained here: https://github.com/sbt/sbt/issues/2786#issuecomment-261773807

**Comment by olafurpg** [comment]
With that approach, the plugin could be implemented almost as easily as
```scala
import sbt._, Keys._

object ScalafmtPlugin extends AutoPlugin {
  override def extraProjects: Seq[Project] =
    List("scalafmt") map generateProject

  def generateProject(id: String): Project =
    Project(id, file(id)).
      settings(
        name := id,
        scalaVersion := "2.11.8",
        libraryDependencies += "com.geirsson" % "scalafmt-cli_2.11" % "0.4.10"
      )
}
```

**Comment by RomanIakovlev** [comment]
I'm probably missing something obvious, but I haven't worked with synthetic projects in sbt 0.13.13 yet, so I don't see how they'll help in this situation. I think I'll try the coursier-based approach first, but will look at synthetic projects after I'll have a proof of concept running.

**Comment by clhodapp** [comment]
If anyone wants to riff on my hack at https://github.com/clhodapp/bijection/blob/master/project/ScalaFmt.scala, feel free. The output is still a little janky, but it seems to work. I went with the generated project idea.

**Comment by clhodapp** [comment]
To explain what's going on more fully: I'm setting up a synthetic stub project which just pulls down the scalafmt CLI. I've set it up so that the `run` task for that project invokes the `main` method for the CLI and then set up a task on each project that will invoke `run` on the stub project, passing the target project's scala sources as arguments.

**Comment by olafurpg** [comment]
@clhodapp Cool! I'm glad to hear that approach seems to work, thanks for sharing. Do you often run `scalafmt` from a subproject instead of all projects in the build?

If not, you might get away with some of the `unmanagedSources` complexity by not passing in the `--files` flag and using instead the `project.git = true` setting in .scalafmt.conf. For more details, see https://olafurpg.github.io/scalafmt/#project

**Comment by olafurpg** [comment]
I decided to add an sbt plugin using the synthetic project idea. Using this approach, the plugin is just a tiny wrapper around the command line interface and should therefore have a lower maintenance cost. For detailed comparison: #610

**Comment by olafurpg** [comment]
0.5 was released with a new sbt plugin merged in #610.
